src/backend/d3d12/ResourceUploader.h

```diff
@@ -0,0 +1,50 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BACKEND_D3D12_RESOURCEUPLOADER_H_
+#define BACKEND_D3D12_RESOURCEUPLOADER_H_
+
+#include "d3d12_platform.h"
+
+#include <map>
+#include <vector>
+
+namespace backend {
+namespace d3d12 {
+
+    class Device;
+
+    class ResourceUploader {
+        public:
+            ResourceUploader(Device* device);
+
+            void UploadResource(ComPtr<ID3D12Resource> resource, uint32_t start, uint32_t count, const uint32_t* data);
+            bool HasPendingResources() const;
+            void SignalUploadStart();
+            void FreeCompletedResources();
+
+        private:
+            Device* device;
+
+            ComPtr<ID3D12Fence> fence;
+            uint64_t uploadSerial = 0;
+            std::vector<ComPtr<ID3D12Resource>> pendingResources;
+            std::multimap<uint64_t, ComPtr<ID3D12Resource>> uploadMap;
```

Comment:
Things tracking what's in flight could be std::vector<std::pair<serial, std::vector<ThingInFlight>>>.